src/clj/puppetlabs/jdbc_util/cp.clj

```diff
@@ -0,0 +1,183 @@
+   :connection-test-query    "/* hikaricp test query */ select 1;"
+   :isolate-internal-queries true
+   ;; Statistics default settings
+   :stats                    true
+   :metrics-registry         metrics/default-registry})
+
+;; Schema
+
+(def pooled-db-spec-schema
+  "Schema representing a clojure.java.jdbc compatible pooled db-spec."
+  {:datasource javax.sql.DataSource})
+
+(defn- gte-0?
+  "Returns true if num is greater than or equal 0, else false"
+  [x]
+  (>= x 0))
+
+(defn- gte-1?
+  "Returns true if num is greater than or equal 1, else false"
+  [x]
+  (>= x 1))
+
+(defn- gte-1000?
+  "Returns true if num is greater than or equal 1000, else false"
+  [x]
+  (>= x 1000))
+
+(def ^{:private true} IntGte0
+  (s/both s/Int (s/pred gte-0? 'gte-0?)))
+
+(def ^{:private true} IntGte1
+  (s/both s/Int (s/pred gte-1? 'gte-1?)))
+
+(def ^{:private true} IntGte1000
+  (s/both s/Int (s/pred gte-1000? 'gte-1000?)))
+
+(def ConfigurationOptions
+  {:auto-commit                         s/Bool
+   (s/optional-key :classname)          s/Str
+   (s/optional-key :connection-timeout) IntGte1000
+   (s/optional-key :idle-timeout)       IntGte0
+   :initialization-fail-fast            s/Bool
+   (s/optional-key :max-lifetime)       IntGte0
+   (s/optional-key :maximum-pool-size)  IntGte1
+   :metrics-registry                    MetricRegistry
+   (s/optional-key :minimum-idle)       IntGte0
+   (s/optional-key :password)           s/Str
+   :pool-name                           s/Str
+   (s/optional-key :read-only)          (s/maybe s/Bool)
+   :register-mbeans                     s/Bool
+   :stats                               s/Bool
+   :subname                             s/Str
+   :subprotocol                         s/Str
+   (s/optional-key :validation-timeout) IntGte1000
+   (s/optional-key :user)               s/Str
+   :connection-test-query               s/Str
+   :isolate-internal-queries            s/Bool})
+
+;; Functions
+
+(s/defn ^:always-validate add-datasource-property
+  "Add a custom datasource property to the underlying database driver."
+  [config :- HikariConfig
+   property :- s/Str
+   value :- s/Str]
+  (.addDataSourceProperty config property value))
+
+(s/defn ^:always-validate add-datasource-properties
+  "Like add-datasource-property, but allows you to pass a map of properties and
+   values."
+  [config :- HikariConfig
+   properties :- {s/Str s/Str}]
+  (doseq [[k v] properties]
+    (add-datasource-property config k v)))
+
+(defn validate-options
+  "Manage defaulting and validation of options."
+  [options]
+  (s/validate ConfigurationOptions (merge default-datasource-options options)))
+
+(s/defn ^:always-validate datasource-config :- HikariConfig
+  "Produce a valid HikariConfig object from provided configuration options."
+  [datasource-options]
+  (let [config (HikariConfig.)
+        options (validate-options datasource-options)
+        {:keys [adapter
```

Comment:
This is all very verbose and I was going to come back to it and condense some of it later.